### ISSUE TITLE
Add the ability to forward props to slots

### ISSUE THREAD

**Issue by haikyuu** [imba/imba#752]
This feature would be very useful and is quite common in vue/svelte
```
tag parent
  <self> <slot name="my-child">

tag kid
  <self> name

tag app
  <self> <parent> <kid>
```
the `kid` tag should receive the name prop from the parent.

**Comment by familyfriendlymikey** [comment]
I don't use slots often but it seems interesting. Though I would think that `slot` can just be implicit, as in:
```imba
tag my-component
  <self>
    <slot text="greetingMessage" count=1>

tag app
  <self>
    <my-component>
      slot.text + slot.count
```

**Comment by andregoldstein** [comment]
This is something I'd find really useful too but not sure how tricky it is to accomplish architecturally.

**Comment by haikyuu** [comment]
implicit is tricky, what about nested slots?
```
<my-component>
  <my-component>
    slot.text # which one would it be? how to use the other?
```
I have the feeling that the render prop is the best:
- it reads well
- doesn't add indentation
- you can name the prop as you like
- no problem with nested slots

**Comment by IRod22** [comment]
> implicit is tricky, what about nested slots?
> 
> ```
> <my-component>
>   <my-component>
>     slot.text # which one would it be? how to use the other?
> ```
> 
> I have the feeling that the render prop is the best:
> 
> ```
> tag app
>   <self>
>     <my-component> do(slotProps)
>        slotProps.text + slotProps.count
> ```
> 
>     * it reads well
> 
>     * doesn't add indentation
> 
>     * you can name the prop as you like
> 
>     * no problem with nested slots

@haikyuu I like that idea, but there would have to be a way to implement a sort of `props.children` for Imba, or implicit props would need to be implemented like in Vue or Svelte.

**Comment by IRod22** [comment]
On a separate note, I would like to say that although Imba is barely in alpha testing, it looks like the future of full stack development; also, I would be glad to give some input to help out this project.

**Comment by haikyuu** [comment]
> On a separate note, I would like to say that although Imba is barely in alpha testing, it looks like the future of full stack development; also, I would be glad to give some input to help out this project.

Imba is already used in production in scrimba.com and some other sites.

The main reason behind it being in alpha is  API stability. There are some new features that are WIP that we want to include in a stable 2.0 release.

**Comment by IRod22** [comment]
I have done some research, and with my understanding of how Imba works, implementing the render function example with named slots will not work with the current implementation because the component will have to dig into its child content for the slots to work; in addition, the same example will not be able to work with custom elements because neither the Web Components API nor the DOM support functions as child nodes.

There are three alternate options available: either there is a `slot[:name][="props"]` like in Vue, there is a `[slot="name"] [... let:alias[={prop}]]` like in Svelte, or we go full React and ditch the `slot` attribute in favor for full render props but at the cost of losing custom elements, [which is how SolidJS works](https://www.solidjs.com/tutorial/flow_show). Here are examples of each:

#### Vue and Svelte slots in Imba
```imba
tag Card
    <self> <.card>
        <.card-content> <slot>
        <.card-description> <slot name="description">
tag App
    <self> <Card>
        "Card content."
        <span slot="description"> "Here is the description."
```
#### React render props in Imba
```imba
tag Card
    prop children
    prop description
    <self> <.card>
        <.card-content> children
        <.card-description> description!
tag App
    <self> <Card description=(do
        <span> "Here is the description.")
    > "Card content."
```

**Comment by haikyuu** [comment]
@IRod22 Named slots already work today in imba.

And render props work as well*. Here is the [example from React docs](https://beta.reactjs.org/apis/react/Children#calling-a-render-prop-to-customize-rendering) in imba https://scrimba.com/scrim/cG8363TN

```imba
tag tab-switcher
	prop tabIds 
	prop selectedId
	prop renderContent
	prop getHeader
	def setup
		selectedId = tabIds[0]
	<self>
		for tabId in tabIds
			<button @click=(selectedId = tabId)>
				getHeader(tabId)
		<hr>
		<div>
			<h3> getHeader(selectedId)
			renderContent(selectedId)

tag App
	def getHeader(tabId)
		tabId[0].toUpperCase() + tabId.slice(1)
	def renderContent(tabId)
		<p> "This is the {tabId} item"
	<self>
		<tab-switcher tabIds=["first", "second", "third"] getHeader=getHeader renderContent=renderContent>

imba.mount <App>
```
The variant of render props that I'm referring to here is the one in use with React's `children`, like this
```imba
<parent> do(name)
    <h1> name
```

**Comment by haikyuu** [comment]
From the implementation side, I think we can convert
```
tag my-tag
  <self> 
    <tab-switcher> do(title) 
      <h1> title
```
to 
```
tag my-tag
  def __slot_prop__(title)
    <h1> title
  <self>
    <tab-switcher __slot_prop__=__slot_prop__>
```
And the `tab-switcher` tag would consume it like this
```
tag tab-switcher
  tab-title = "My title"
  <self>
    slot(tab-title)
```

This is a transformation that would be done when parsing imba tags. And we'd make slot callable

**Comment by IRod22** [comment]
> From the implementation side, I think we can convert
> 
> ```
> tag my-tag
>   <self> 
>     <tab-switcher> do(title) 
>       <h1> title
> ```
> 
> to
> 
> ```
> tag my-tag
>   def __slot_prop__(title)
>     <h1> title
>   <self>
>     <tab-switcher __slot_prop__=__slot_prop__>
> ```
> 
> And the `tab-switcher` tag would consume it like this
> 
> ```
> tag tab-switcher
>   tab-title = "My title"
>   <self>
>     slot(tab-title)
> ```
> 
> This is a transformation that would be done when parsing imba tags. And we'd make slot callable

Good idea, @haikyuu, but I was trying to say that we can keep the `slot` tag and attributes if we incorporate a `let:prop`/`let:alias=prop` syntax:

#### inside `<parent>`
```
<slot name="my-slot" my-prop=some-value>
```

#### consuming `<parent>`
```
<parent>
    <div slot="my-slot" let:alias=my-prop> "..."
```

That is how [Svelte](https://svelte.dev/tutorial/slot-props) implements content projection.

**Comment by IRod22** [comment]
I looked at the NPM trends for the [mainstream libraries/frameworks](https://npmtrends.com/@angular/core-vs-react-vs-solid-vs-svelte-vs-vue), and I found that React takes the lead. The trends show that the developer community most likely likes the flexibility of React's JSX syntax. An article [comparing HTML templates and JSX](https://codeburst.io/jsx-and-html-templates-893ca24c1987) also gives a good point about the efficiency of JSX. Although I am open to other suggestions, with that evidence, I can say that Imba may need to lean more towards React regarding slot props.
